fix(pydantic): Fix pydantic default  factory conversion

thejaminator opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

Fixes issue when using a pydantic.Field(default_factory = something)

My bad, forgot to add a test for that

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to James Chua for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

thejaminator commented on strawberry/experimental/pydantic/utils.py:
```diff
@@ -85,9 +85,11 @@ def defaults_into_factory(
+            if default is not None and default_factory is not None:
+                raise BothDefaultAndDefaultFactoryDefinedError(
+                    default=default, default_factory=default_factory  # type: ignore
+                )
```
hacky?

if you looked at pydantic's ModelField, it sets default and default_factory to None by default.
rather than having a value such as UNSET to differentiate between None as the user's default, and having no default.

thejaminator commented on tests/experimental/pydantic/test_basic.py:
```diff
@@ -272,6 +272,46 @@ class UserType:
+
+    class User3(pydantic.BaseModel):
+        friend: Optional[str] = pydantic.Field(default_factory=lambda: "friend")
+
```
previously defining it like this would fail, because default=None, and we'll think that a factory and default was both defined. 
see https://discord.com/channels/689806334337482765/773519351423827978/927667160250212462

thejaminator commented on tests/experimental/pydantic/test_conversion.py:
```diff
@@ -787,22 +787,6 @@ def test_convert_input_types_to_pydantic_default_and_default_factory():
-        @strawberry.experimental.pydantic.input(User2)
-        class UserInput:
-            password: strawberry.auto
-
```
turns out hacking it like this wasn't possible. 
doing
```
hacked_field.default = None 
```
didn't actually change the default. pydantic under the hood would ignore the change somehow.

the test only worked because it already was None by default

thejaminator commented on strawberry/experimental/pydantic/object_type.py:
```diff
@@ -121,7 +121,10 @@ def wrap(cls):
+                        # pydantic's default and default_factory are None by default
+                        default=field.default
+                        if (not field.required) and (field.default_factory is None)
+                        else UNSET,
```
turns out pydantic's `ModelField` doesn't use it's Undefined class to mark that the default was undefined. Thats only used in `Field`, when the user defines it. Once the model class is actually instantiated, `Field` gets turned into a `ModelField` and that Undefined distinction gets lost.

thejaminator replied:
since field.default_factory is `Optional[NoArgAnyCallable]`, if it is None and if the field is not required, 
then the default value must come from field.default

patrick91 commented:
@thejaminator I did change this a bit, would you be able to check if the changes make sense? 

Also let me know if you want a pre-release if you have some code to test this 😊

thejaminator commented:
> @thejaminator I did change this a bit, would you be able to check if the changes make sense?
> 
> Also let me know if you want a pre-release if you have some code to test this 😊

@patrick91  i added a few test statements and corrected a typo. looks good! i tested it locally on my codebase and seems fine, so won't need a pre-release :D

patrick91 commented:
Awesome! Thanks ☺️
